baseplate/lib/metrics.py

```diff
@@ -112,14 +112,20 @@ def flush(self) -> None:
 class RawTransport(Transport):
     """A transport which sends messages on a socket."""
 
-    def __init__(self, endpoint: config.EndpointConfiguration):
+    def __init__(
+        self, endpoint: config.EndpointConfiguration, swallow_network_errors: bool = False,
+    ):
+        self.swallow_network_errors = swallow_network_errors
         self.socket = socket.socket(endpoint.family, socket.SOCK_DGRAM)
         self.socket.connect(endpoint.address)
 
     def send(self, serialized_metric: bytes) -> None:
         try:
             self.socket.sendall(serialized_metric)
         except socket.error as exc:
+            if self.swallow_network_errors:
+                logger.exception("Failed to send to metrics collector")
+                return
             if exc.errno == errno.EMSGSIZE:
                 raise MessageTooBigTransportError(len(serialized_metric))
```

Comment:
I'd hate to lose the special error messaging around this one as it's something people have asked a lot about in the past. Since we catch-and-swallow this exception when we flush a batch, could the swallowing check here come after this re-raise? One small risk is we might get an EMSGSIZE false positive in a sidecar and it'd still cause a crash, but I think that's generally pretty infrequent?

https://github.com/reddit/baseplate.py/blob/729452d6246332cd894cc4cc7107b7ff51615531/baseplate/lib/metrics.py#L251

Reply:
I think that should be safe enough for sidecar use case, but it kind of contradicts to the "swallow_network_errors" name as when people uses that config in other environment they might expect these errors are also swallow?

I don't feel strongly either way.

Reply:
but I do agree that the error message of that exception would be very useful, so if we just add another layer of try...catch (but only catch on `MessageTooBigTransportError` and `TransportError`), do you consider that too ugly? 😅

Reply:
hrm, yeah, true. I think the only times this error should legitimately happen would be from using a batch and putting too much in it, in which case the error would still be swallowed. But if we really do get false positives with any frequency then it kinda sucks to not swallow it, like you say. Let's go with not swallowing that one at this layer just because I think it's far more common to have the batch case happen and that's caused a lot of support requests so having the extra messaging is helpful.

Reply:
Do you like https://github.com/reddit/baseplate.py/pull/576/commits/92e0f900420962fc6a4ba3c49ae01503aea25bae better?

Reply:
hmm. the more I think about it, I think the commit before was fine. We should generally only be doing the error swallowing in sidecars, not main services, and services are where the special error messaging matters. Going back to that OK? sorry :(

Reply:
probably also want to hard-code swallowing network errors when setting up the metrics client in sidecars? so we don't even need to make everyone change configs https://github.com/reddit/baseplate.py/blob/develop/baseplate/sidecars/trace_publisher.py#L170

Reply:
I... am not sure that's feasible to do.

So to do that I think the best way is to duplicate `metrics_client_from_config` into another version `metrics_client_from_config_for_sidecars` and remove `swallow_network_errors` from the config (and hardcode it in `metrics_client_from_config_for_sidecars`)? Otherwise we would need to get a client from config first, then access its transport to change the `swallow_network_errors` field to True.